internal/controller/serverset/serverset_crossplane_mock_test.go

```diff
@@ -0,0 +1,242 @@
+				},
+				log: logging.NewNopLogger(),
+				mg:  &fake.Managed{},
+				bootVolumeController: &kubeBootVolumeController{
+					kube: &test.MockClient{
+						MockCreate: test.NewMockCreateFn(nil),
+						MockList:   test.NewMockListFn(nil, fakeObjList),
+						MockGet: test.NewMockGetFn(nil, func(obj client.Object) error {
+							volume := obj.(*v1alpha1.Volume)
+							annotations := make(map[string]string)
+							annotations[meta.AnnotationKeyExternalCreateFailed] = "alas, we failed"
+							volume.SetAnnotations(annotations)
+							return nil
+						}),
+						MockDelete: test.NewMockDeleteFn(nil),
+					},
+					log: logging.NewNopLogger(),
+				},
+			},
+			args: args{
+				ctx: context.Background(),
+				cr:  createBasicServerSet(),
+			},
+			wantErr: fmt.Errorf("while waiting for BootVolume to be populated %w ", kube.ErrExternalCreateFailed),
+		},
+		{
+			name: "bad cloud init",
+			fields: fields{
+				kube: &test.MockClient{
+					MockList: test.NewMockListFn(nil, fakeObjList),
+				},
+				log: logging.NewNopLogger(),
+				mg:  &fake.Managed{},
+				bootVolumeController: &kubeBootVolumeController{
+					kube: &test.MockClient{
+						MockCreate: test.NewMockCreateFn(nil),
+						MockList:   test.NewMockListFn(nil, fakeObjList),
+						MockGet: test.NewMockGetFn(nil, func(obj client.Object) error {
+							volume := obj.(*v1alpha1.Volume)
+							annotations := make(map[string]string)
+							annotations[meta.AnnotationKeyExternalCreateFailed] = "alas, we failed"
+							volume.SetAnnotations(annotations)
+							return nil
+						}),
+						MockDelete: test.NewMockDeleteFn(nil),
+					},
+					log: logging.NewNopLogger(),
+				},
+			},
+			args: args{
+				ctx: context.Background(),
+				cr: &v1alpha1.ServerSet{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: serverSetName,
+						Annotations: map[string]string{
+							"crossplane.io/external-name": serverSetName,
+						},
+					},
+					Spec: v1alpha1.ServerSetSpec{
+						ForProvider: v1alpha1.ServerSetParameters{
+							Replicas: noReplicas,
+							Template: v1alpha1.ServerSetTemplate{
+								Metadata: v1alpha1.ServerSetMetadata{
+									Name: "servername",
+								},
+								Spec: v1alpha1.ServerSetTemplateSpec{
+									Cores:     serverSetCores,
+									RAM:       serverSetRAM,
+									CPUFamily: serverSetCPUFamily,
+									NICs: []v1alpha1.ServerSetTemplateNIC{
+										{
+											Name:      "nic1",
+											IPv4:      "10.0.0.2/24",
+											DHCP:      true,
+											Reference: "data",
+										},
+									},
+								},
+							},
+							BootVolumeTemplate: v1alpha1.BootVolumeTemplate{
+								Metadata: v1alpha1.ServerSetBootVolumeMetadata{
+									Name: "bootvolumename",
+								},
+								Spec: v1alpha1.ServerSetBootVolumeSpec{
+									Size:     bootVolumeSize,
+									Image:    bootVolumeImage,
+									Type:     bootVolumeType,
+									UserData: "malformed",
+								},
+							},
+						},
+					},
+				},
```

Comment:
Why didn't you create a new method which extends the functionality from `createBasicServerSet()` by populating the UserData field?

Reply:
Will do it when I see that it can actually be used in other places